src/hotspot/os/posix/signals_posix.cpp

```diff
@@ -502,21 +528,106 @@ void PosixSignals::unblock_error_signals() {
+  // Note: it's not uncommon that JNI code uses signal/sigset to install,
+  // then restore certain signal handler (e.g. to temporarily block SIGPIPE,
+  // or have a SIGILL handler when detecting CPU type). When that happens,
+  // this handler might be invoked with junk info/ucVoid. To avoid unnecessary
+  // crash when libjsig is not preloaded, try handle signals that do not require
+  // siginfo/ucontext first.
+
+  // Preserve errno value over signal handler.
+  //  (note: RAII ok here, even with JFR thread crash protection, see below).
+  ErrnoPreserver ep;
+
+  // Unblock all synchronous error signals (see JDK-8252533)
   PosixSignals::unblock_error_signals();
 
-  int orig_errno = errno;  // Preserve errno value over signal handler.
-#if defined(BSD)
-  JVM_handle_bsd_signal(sig, info, uc, true);
-#elif defined(AIX)
-  JVM_handle_aix_signal(sig, info, uc, true);
+  ucontext_t* const uc = (ucontext_t*) ucVoid;
+  Thread* const t = Thread::current_or_null_safe();
+  JavaThread* const jt = (t != NULL && t->is_Java_thread()) ? (JavaThread*) t : NULL;
+
+  // Handle JFR thread crash protection.
+  //  Note: this may cause us to longjmp away. Do not use any code before this
+  //  point which really needs any form of epilogue code running, eg RAII objects.
+  os::ThreadCrashProtection::check_crash_protection(sig, t);
+
+  bool signal_was_handled = false;
+
+  // Handle assertion poison page accesses.
+#ifdef CAN_SHOW_REGISTERS_ON_ASSERT
+  if ((sig == SIGSEGV || sig == SIGBUS) && info != NULL && info->si_addr == g_assert_poison) {
+    signal_was_handled = handle_assert_poison_fault(ucVoid, info->si_addr);
+  }
+#endif
+
+  // Ignore SIGPIPE and SIGXFSZ (4229104, 6499219).
+  if (sig == SIGPIPE || sig == SIGXFSZ) {
+    PosixSignals::chained_handler(sig, info, ucVoid);
+    signal_was_handled = true; // unconditionally.
+  }
+
+  // Call platform dependent signal handler.
+  if (!signal_was_handled) {
+    signal_was_handled = PosixSignals::pd_hotspot_signal_handler(sig, info, uc, jt);
+  }
+
+  // From here on, if the signal had not been handled, it is a fatal error.
+
+  // Give the chained signal handler - should it exist - a shot.
+  if (!signal_was_handled) {
+    signal_was_handled = PosixSignals::chained_handler(sig, info, ucVoid);
+  }
+
+  // Invoke fatal error handling.
+  if (!signal_was_handled && abort_if_unrecognized) {
+    // Extract pc from context for the error handler to display.
+    address pc = NULL;
+    if (uc != NULL) {
+      // prepare fault pc address for error reporting.
+      if (S390_ONLY(sig == SIGILL || sig == SIGFPE) NOT_S390(false)) {
+        pc = (address)info->si_addr;
+      } else {
+        pc = PosixSignals::ucontext_get_pc(uc);
+      }
+    }
+#if defined(ZERO) && !defined(PRODUCT)
+    char buf[20];
+    VMError::report_and_die(t, sig, pc, info, ucVoid,
+          "\n#"
+          "\n#    /--------------------\\"
+          "\n#    |      %-7s       |"
+          "\n#    \\---\\ /--------------/"
```

Comment:
Isn't the little robot supposed to say "segmentation fault" and would that be safe than calling get_signal_name in this context?

Reply:
Thanks for keeping the picture.

Reply:
Thanks Coleen. Little robot now spells out the name of the signal (since "segmentation fault" is only correct for segv, and I wanted to see it for the other cases too. get_signal_name() is completely harmless, just uses a bit of stack buffer (and that only for the case of unknown signals which are printed numerically; I plan to change that and give us a simple version which just returns static strings).

Reply:
OK.